cdap-common/src/main/java/co/cask/cdap/internal/app/store/RunRecordMeta.java

```diff
@@ -186,7 +227,7 @@ public RunRecordMeta build() {
         throw new IllegalArgumentException("Run record source id must be specified.");
       }
       return new RunRecordMeta(programRunId, startTs, runTs, stopTs, status, properties, systemArgs, twillRunId,
-                               cluster, sourceId);
+                               cluster, sourceId, artifactId, principal);
```

Comment:
should validate that artifactId is not null

Reply:
yes, but if there were some notifications in the topic about program start just before the upgrade to the new version, and after upgrade while processing those notifications artifactId will be null. if we throw exception for them, then we wouldn't record those state changes.

Reply:
ah I see. Can you add a comment then, explaining why it is nullable?